Issues with Galen Testsuite and Selenium Grid Args

When using the Galen TestSuite syntax there seem some issues, that the handover of the desired capabilities to Galen is not working from CLI to Java, see [here](https://github.com/hypery2k/galenframework-cli/issues/43).

I will try to add a test case in my fork and also a PR. Just in case anybody else facing this issue

@ishubin 

I've checked the setup at customer side and it's kind of weird. Chrome works, but Firefox and MSIE are not working for 2 of 4 sessions:

```
========================================
Test: Survey on Phone (MSIE) (S) @ develop
========================================
java.lang.RuntimeException: org.openqa.selenium.SessionNotCreatedException: Unable to create new remote session. desired capabilities = Capabilities [{browserName=internet explorer}], required capabilities = Capabilities [{}]
Build info: version: '3.2.0', revision: '8c03df6', time: '2017-03-02 09:34:51 -0800'
System info: host: 'SVMOXF0101', ip: '10.79.78.163', os.name: 'Windows 10', os.arch: 'x86', os.version: '10.0', java.version: '1.8.0_112'
Driver info: driver.version: InternetExplorerDriver
Command duration or timeout: 0 milliseconds
	at com.galenframework.browser.SeleniumGridBrowserFactory.openBrowser(SeleniumGridBrowserFactory.java:70)
	at com.galenframework.runner.GalenBasicTestRunner.runTest(GalenBasicTestRunner.java:73)
	at com.galenframework.tests.GalenBasicTest.execute(GalenBasicTest.java:55)
	at com.galenframework.TestRunnable.runTest(TestRunnable.java:75)
	at com.galenframework.TestRunnable.run(TestRunnable.java:101)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.openqa.selenium.SessionNotCreatedException: Unable to create new remote session. desired capabilities = Capabilities [{browserName=internet explorer}], required capabilities = Capabilities [{}]
Build info: version: '3.2.0', revision: '8c03df6', time: '2017-03-02 09:34:51 -0800'
System info: host: 'SVMOXF0101', ip: '10.79.78.163', os.name: 'Windows 10', os.arch: 'x86', os.version: '10.0', java.version: '1.8.0_112'
Driver info: driver.version: InternetExplorerDriver
Command duration or timeout: 0 milliseconds
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:216)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:168)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:367)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:113)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:141)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:604)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:244)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:131)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:158)
	at com.galenframework.browser.SeleniumGridBrowserFactory.openBrowser(SeleniumGridBrowserFactory.java:65)
	... 7 more
Caused by: org.openqa.selenium.SessionNotCreatedException: Unable to create new remote session. desired capabilities = Capabilities [{browserName=internet explorer}], required capabilities = Capabilities [{}]
Build info: version: '3.2.0', revision: '8c03df6', time: '2017-03-02 09:34:51 -0800'
System info: host: 'SVMOXF0101', ip: '10.79.78.163', os.name: 'Windows 10', os.arch: 'x86', os.version: '10.0', java.version: '1.8.0_112'
Driver info: driver.version: InternetExplorerDriver
Build info: version: 'unknown', revision: 'unknown', time: 'unknown'
System info: host: 'fabi-XPS-15-9550', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-78-generic', java.version: '1.8.0_121'
Driver info: driver.version: InternetExplorerDriver
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:91)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:141)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:82)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:604)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:244)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:230)
	at org.openqa.selenium.ie.InternetExplorerDriver.run(InternetExplorerDriver.java:180)
	at org.openqa.selenium.ie.InternetExplorerDriver.<init>(InternetExplorerDriver.java:172)
	at org.openqa.selenium.ie.InternetExplorerDriver.<init>(InternetExplorerDriver.java:148)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.lang.reflect.Constructor.newInstance(Unknown Source)
	at org.openqa.selenium.remote.server.DefaultDriverProvider.callConstructor(DefaultDriverProvider.java:103)
	at org.openqa.selenium.remote.server.DefaultDriverProvider.newInstance(DefaultDriverProvider.java:97)
	at org.openqa.selenium.remote.server.DefaultDriverFactory.newInstance(DefaultDriverFactory.java:60)
	at org.openqa.selenium.remote.server.DefaultSession$BrowserCreator.call(DefaultSession.java:222)
	at org.openqa.selenium.remote.server.DefaultSession$BrowserCreator.call(DefaultSession.java:209)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at org.openqa.selenium.remote.server.DefaultSession$1.run(DefaultSession.java:176)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)

```

--- a/galen-core/src/main/java/com/galenframework/browser/SeleniumGridBrowserFactory.java
+++ b/galen-core/src/main/java/com/galenframework/browser/SeleniumGridBrowserFactory.java
@@ -42,33 +42,42 @@ public class SeleniumGridBrowserFactory implements BrowserFactory {
 
     @Override
     public Browser openBrowser() {
-        DesiredCapabilities desiredCapabilities = new DesiredCapabilities();
-        
-        if (platform != null) {
-            desiredCapabilities.setPlatform(platform);
-        }
-        
-        if (browser != null) {
-            desiredCapabilities.setBrowserName(browser);
-        }
-        
-        if (browserVersion != null) {
-            desiredCapabilities.setVersion(browserVersion);
-        }
-        
-        for (Map.Entry<String, String> dc : this.desiredCapabilities.entrySet()) {
-            desiredCapabilities.setCapability(dc.getKey(), dc.getValue());
-        }
         
         try {
             
-            WebDriver driver = new RemoteWebDriver(new URL(gridUrl), desiredCapabilities);
+            WebDriver driver = new RemoteWebDriver(new URL(gridUrl), this.createCapabilities());
             WebDriver augmentedDriver = new Augmenter().augment(driver);
         	return new SeleniumBrowser(augmentedDriver);
         }
         catch (Exception ex) {
         	throw new RuntimeException(ex);
         }
+    }
+
+    public DesiredCapabilities createCapabilities(){
+        DesiredCapabilities desiredCapabilities = new DesiredCapabilities();
+
+        if (platform != null) {
+            desiredCapabilities.setPlatform(platform);
+        }
+
+        if (browser != null) {
+            desiredCapabilities.setBrowserName(browser);
+        }
+
+        if (browserVersion != null) {
+            desiredCapabilities.setVersion(browserVersion);
+        }
+
+        for (Map.Entry<String, String> dc : this.desiredCapabilities.entrySet()) {
+            final String value = dc.getValue();
+            if("true".equals(value) || "false".equals(value)) {
+                desiredCapabilities.setCapability(dc.getKey(), Boolean.parseBoolean(value));
+            } else {
+                desiredCapabilities.setCapability(dc.getKey(), value);
+            }
+        }
+        return desiredCapabilities;
     }
 
     public SeleniumGridBrowserFactory withBrowser(String browser) {

--- a/galen-core/src/test/java/com/galenframework/tests/browser/SeleniumGridBrowserFactoryTest.java
+++ b/galen-core/src/test/java/com/galenframework/tests/browser/SeleniumGridBrowserFactoryTest.java
@@ -1,0 +1,47 @@
+/*******************************************************************************
+ * Copyright 2017 Ivan Shubin http://galenframework.com
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+package com.galenframework.tests.browser;
+
+import static org.testng.AssertJUnit.assertTrue;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.openqa.selenium.remote.DesiredCapabilities;
+import org.testng.annotations.Test;
+
+import com.galenframework.browser.SeleniumGridBrowserFactory;
+
+public class SeleniumGridBrowserFactoryTest {
+
+    // See https://github.com/galenframework/galen/issues/494
+    @Test
+    public void shouldParseStringAndBooleanGridArgsCorrectly() {
+        // given
+        Map<String, String> inputCaps = new HashMap<>();
+        inputCaps.put("marionette", "true");
+        inputCaps.put("anString", "42");
+        // when
+        SeleniumGridBrowserFactory factory = new SeleniumGridBrowserFactory("http://mygrid.com");
+        factory.setBrowser("firefox");
+        factory.setDesiredCapabilites(inputCaps);
+        DesiredCapabilities usedCapsForWebDriver = factory.createCapabilities();
+        // then
+        assertTrue((Boolean) usedCapsForWebDriver.getCapability("marionette"));
+        assertTrue(usedCapsForWebDriver.getBrowserName().equalsIgnoreCase("firefox"));
+    }
+}

